File: 剑指offer刷题/剑指Offer-JavaScript实现/7_Fibonacci.js
/* 
7. 斐波那契数列
题目描述
大家都知道斐波那契数列，现在要求输入一个整数n，请你输出斐波那契数列的第n项（从0开始，第0项为0，第1项是1）。
n<=39 
*/

// 递归
function Fibonacci(n)
{
    // write code here
    if(n<1){
        return 0
    }
    if(n == 1 || n == 2){
        return 1
    }
    return Fibonacci(n-2) + Fibonacci(n-1)
}
// 尾递归优化
function Fibonacci_2(n, acc, cal){
    if (n<1) {
        return 0
    }
    if (n==1) {
        return acc
    }
    if (n==2) {
        return cal
    }   
    return Fibonacci_2(n-1, cal, acc+cal)
}
// 使用额外空间
function Fibonacci_3(n){
    if(n<1){
        return 0
    }
    let a = 0
    let b = 1
    let index = 1
    while (index<n) {
        let temp = a
        a = b
        b = a + temp
        index += 1
    }
    return b
}

// 测试用例
var ans = Fibonacci(9)
console.log(ans)

for (let index = 0; index < 10; index++) {
    const element = Fibonacci(index);
    console.log(index,element)  
}

var ans = Fibonacci_3(89)
console.log(ans)

File: SwordToOffer/SwordToOffer-PythonSolution/7_fibonacci.py
# -*- coding:utf-8 -*-
class Solution:
    
#    使用递归
    def Fibonacci_1(self, n):
        # write code here
        if(n<1):
            return 0
        if(n == 1 or n == 2):
            return 1
        return self.Fibonacci_1(n-1)+self.Fibonacci_1(n-2)
    
    # 优化递归
    def Fibonacci_2(self, n):
        # write code here
        if(n<1):
            return 0
        a, b = 0, 1
        i = 1
        while i<n:
            a,b = b,a+b
            i += 1
        return b
    
    # 尾递归优化 效果同上种方法
    def Fibonacci_3(self, n, acc, cal):
        # write code here
        if(n<1):
            return 0
        if(n == 1):
            return acc
        if(n == 2):
            return cal
        return self.Fibonacci_3(n-1,cal,acc+cal)
    
    # 使用额外空间
    def Fibonacci_4(self,n):
        if(n<1):
            return 0
        res = [1,1]
        while len(res)<n:
            res.append(res[len(res)-2]+res[len(res)-1])
        return res[n-1]
        
    
if __name__=='__main__':
    solution = Solution()
    for i in range(7):
        ans = solution.Fibonacci_4(i)
        print('ans=',ans)